src/EditorFeatures/TestUtilities/StubVsServiceExporter`2.cs

```diff
@@ -0,0 +1,62 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.ComponentModel.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.VisualStudio;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Threading;
+
+namespace Microsoft.CodeAnalysis.Editor.UnitTests;
+
+[Export(typeof(IVsService<,>))]
+[PartCreationPolicy(CreationPolicy.NonShared)]
+[PartNotDiscoverable]
+internal class StubVsServiceExporter<TService, TInterface> : IVsService<TService, TInterface>
+    where TService : class
+    where TInterface : class
+{
+    private readonly AsyncLazy<TInterface> _serviceGetter;
+    private Task<TInterface?>? _nullableGetter;
+
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public StubVsServiceExporter(
+        [Import(typeof(SAsyncServiceProvider))] IAsyncServiceProvider2 asyncServiceProvider,
+        JoinableTaskContext joinableTaskContext)
+    {
+        _serviceGetter = new AsyncLazy<TInterface>(() => asyncServiceProvider.GetServiceAsync<TService, TInterface>(joinableTaskContext.Factory, throwOnFailure: true)!, joinableTaskContext.Factory);
+    }
+
+    /// <inheritdoc />
+    public Task<TInterface> GetValueAsync(CancellationToken cancellationToken)
+        => _serviceGetter.GetValueAsync(cancellationToken);
+
+    /// <inheritdoc />
+    public Task<TInterface?> GetValueOrNullAsync(CancellationToken cancellationToken)
+    {
+        _nullableGetter ??= _serviceGetter.GetValueAsync(cancellationToken).ContinueWith(
```

Comment:
There's a bug here: if GetValueOrNullAsync() were to be called with a token that is cancelled before we fetch the service, _nullableGetter might be assigned a task that's cancelled. A later call with a different, non-cancelled token would just return that cancelled task for the rest of the lifetime of this object.

Since this is test code, the relatively rare allocations of an extra task here doesn't make it worth trying to be fancy here. This also might be easier though to also write this just as:

```
var task = _serviceGetter.GetValueAsync(cancellaitionToken);
await task.NonThrowable(); // I'm sure we have one of these helpers in scope, not sure which one though
if (task.Stauts == TaskStatus.Faulted) return null;
await task;
```

Reply:
@jasonmalinowski, is there a benefit to doing it the way you suggest rather than the even simpler (IMHO):

```cs
try
{
    return await GetValueAsync(cancellationToken);
}
catch (Exception ex) when (ex is not OperationCanceledException)
{
    return null;
}
```

Reply:
That seems reasonable too!

Reply:
@jasonmalinowski Are you assuming that `AsyncLazy<T>` here comes from Roslyn? Because in this case we're using the vs-threading implementation, and if there's a bug here it means there's a bug in the core implementation of the same feature in Visual Studio itself.

Reply:
➡️ So Jason is correct, but I believe we should retain this behavior because it matches what will happen if this type is used in production code.

Reply:
@sharwell I missed that, but doesn't matter -- the bug exists either way; we'll still cache a potentially cancelled task which is always returned for future calls.

Reply:
(I can see that not happening if it turns out GetValueAsync() from the underlying AsyncLazy won't ever respect cancellation, but at that point this code is all moot trying to handle cancellation. But trying to argue all of that is a lot harder than just taking @jdrobison's code)

Reply:
I had misunderstood @sharwell's comment -- indeed, this appears broken in VS too.

@AArnott while bringing this support into our test harness it appears we've found a bug in the existing VS version. "Let's maintain the same bugs as the real implementation" seems a bit unnecessary since we can fix both! Any preference for a fix to make? I'd say let's just take one of the other implementation's we've discussed in this thread and we can then update the VS version to match. Happy to volunteer to make the PR too.

Reply:
Yes, I see the bug. As you volunteered, yes please fix in both places and include me on the VS PR.

Reply:
I didn't see a PR from you, so I filed one myself: https://dev.azure.com/devdiv/DevDiv/_git/VS/pullrequest/480252